### ISSUE TITLE
Mis-characterization of Frama-C capabilities

### ISSUE THREAD

**Issue by pascal-cuoq** [PolySync/static-analysis-argumentation#1]
Hello,

I suspect that you are misusing the open-source analysis framework Frama-C in your comparison. Any reasonable recent version of Frama-C, used according to its documentation, should report the issues in all your sequential examples. The concurrency example can be analyzed with an analysis plug-in that my employer, [TrustInSoft](https://trust-in-soft.com), provides together with training and support for the entire platform.

Since your sequential examples are deterministic self-contained programs, you can try them out using Firefox or Chrome in our free tsnippet online service.

[This is your alias.c example](https://taas.trust-in-soft.com/tsnippet/t/0ac6fb9d). It does invalid pointer arithmetic (undefined behavior) inside function `black_box`, and ultimately dereferences the invalid address, but [the undefined behavior is creating the invalid address in the first place](https://port70.net/~nsz/c/c11/n1570.html#6.5.6p8).

[This is your constant.c example](https://taas.trust-in-soft.com/tsnippet/t/a29fbd15).  It is still doing invalid pointer arithmetic inside function `black_box`, in a way such that a static analyzer does not even need to detect `const` violations in order to warn about the example, so perhaps you meant something else?

- If you want to consider that converting a pointer to `const t` to a pointer to `t` is bad, then any decent compiler should be able to warn you about it, you do not need advanced static analysis for that.
- If you only want to forbid it when the original pointed object was `const`, and allow it when the original pointed object was not const, that is, if you want to allow line 10 in the following program but forbid line 14, then [Frama-C/TIS Analyzer/tsnippet can be used for that](https://taas.trust-in-soft.com/tsnippet/t/7610c351).

```c
int x = 1;
const int y = 2;

int main(void) {
    const int *p;
    int *q;
    
    p = (const int *)&x;
    q = (int *)p;
    *q = 3;
    
    p = &y;
    q = (int *)p;
    *q = 4;
}
```

[This is your pattern.c example](https://taas.trust-in-soft.com/tsnippet/t/4139bb02). Unfortunately, conversion between integers and enum types are too common in programming practice to forbid them, but Frama-C/TIS Analyzer/tsnippet can warn about the invalid pointer arithmetic and subsequent invalid memory access that happens in this example because of these conversions .

If you have an interest in tools for the verifications of safety-critical C software beyond these examples, I invite you to [contact my employer](https://trust-in-soft.com/contact-us-tis/). Installing a Frama-C package and running it without looking at [what Frama-C claims to provide and how it is used](https://frama-c.com/download/frama-c-value-analysis.pdf) is the equivalent of running a Rust compiler on C programs and reporting that it does not compile them.

**Comment by shnewto** [comment]
Hi there @pascal-cuoq,
Our results are reproducible using Frama-C v16.0 Sulfur. I agree that the TIS tools provide high value reporting but my impression is that they take a program monitoring approach (like Valgrind) rather than strictly static code analysis?

**Comment by pascal-cuoq** [comment]
Yes, your results are reproducible by launching `frama-c program.c`, which makes Frama-C parse the source code and invokes no static analysis on it. So Frama-C remains silent unless program.c contains a syntax error, because this is exactly what has been requested of it.

You seemed to be looking for undefined behavior in your test programs. Since Frama-C's value analysis, invoked  with `-val`, is a mostly automatic analysis to detect undefined behavior, I wanted to suggest you try that. Again, please read [the documentation](https://frama-c.com/download/frama-c-value-analysis.pdf), and you may also want to avoid going to the trouble of making a whitepaper to say the equivalent of “We launched `rustc program.c` and it claimed there was a syntax error. Rust does not seem like a good replacement for C for safety-critical software”.

> I agree that the TIS tools provide high value reporting but my impression is that they take a program monitoring approach (like Valgrind) rather than strictly static code analysis?

Nothing could be further from reality. TIS Analyzer, based on Frama-C, is a sound static analyzer, designed to provide guarantees about all possible behaviors of C programs, even when confronted to billions of possible input values. It is precise enough as a static analyzer that if the program's inputs have been made explicit—say, as the values of global variables—then it can predict exactly what happens during execution and report only problems that happen during execution, without false positives. TrustInSoft provides **free** tools and online services taking advantage of this property because that sub-product is easier to use and requires little training. The free tools and services are not TIS Analyzer, and their limitations are not limitations of TIS Analyzer, even if they reuse some of its building blocks.

**Comment by shnewto** [comment]
Hi @pascal-cuoq,
I agree that there is probably value in exploring the blind spots of other tools and configurations like the Evolved Value Analysis (EVA) plug-in for Frama-C. Enabling EVA does make it easier to spot problematic behavior in examples like ours where every value is defined in the source, but it fails to report any of those problems if, for instance, the values are read from a file. Thanks for the suggestions and feedback.

**Comment by pascal-cuoq** [comment]
Ah, I see that you have closed the ticket. For the record, in case it was not clear, I will summarize the initial issues that I was reporting (and it is your privilege to ignore my reports).

- The whitepaper claims that you have tested Frama-C, but you have not done so in any meaningful sense of the verb. You have not tested one of many “configurations”, but none.

- Your constant.c example is nonsensical: it is doing invalid pointer arithmetic regardless of the `const` qualifier that may or may not be applied erroneously in that program.
